cpp/inc/bond/core/detail/visit_any.h

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#pragma once
+
+#include "mpl.h"
+#include <boost/optional.hpp>
+#include <type_traits>
+
+
+namespace bond
+{
+
+namespace detail
+{
+
+#if !defined(BOND_NO_CXX14_RETURN_TYPE_DEDUCTION) && !defined(BOND_NO_CXX14_GENERIC_LAMBDAS)
+
+template <typename T, typename Visitor, typename Any>
+typename boost::disable_if<std::is_void<std::result_of_t<Visitor(T&)> >, boost::optional<std::result_of_t<Visitor(T&)> > >::type
+try_visit_any(Visitor&& visitor, Any& x)
+{
+    if (auto value = any_cast<T>(&x))
+    {
+        return std::forward<Visitor>(visitor)(*value);
+    }
+
+    return {};
```

Comment:
`boost::none` may be clearer here than `{}`. Below as well for the newer implementation.